htslib/sam.h

```diff
@@ -2101,6 +2134,131 @@ correct thing to do.  It would be wise to avoid this situation if possible.
+                          explicit base recorded in the SEQ column (esp. if N).
+ @field stran             0 or 1, indicating + or - strand from MM tag.
+ @field qual              Quality code (256*probability), or -1 if unknown
+
+ @discussion
+ Note this doesn't hold any location data or information on which other
+ modifications may be possible at this site.
+*/
+typedef struct hts_base_mod {
+    int modified_base;
+    int canonical_base;
+    int strand;
+    int qual;
+} hts_base_mod;
+
+/// Allocates an hts_base_mode_state.
+/**
+ * @return An hts_base_mode_state pointer on success,
+ *         NULL on failure.
+ *
+ * This just allocates the memory.  The initialisation of the contents is
+ * done using bam_parse_basemod.  Successive calls may be made to that
+ * without the need to free and allocate a new state.
+ *
+ * The state be destroyed using the hts_base_mode_state_free function.
+ */
+HTSLIB_EXPORT
+hts_base_mod_state *hts_base_mod_state_alloc(void);
+
+/// Destroys an  hts_base_mode_state.
+/**
+ * @param state    The base modification state pointer.
+ *
+ * The should have previously been created by hts_base_mode_state_alloc.
+ */
+HTSLIB_EXPORT
+void hts_base_mod_state_free(hts_base_mod_state *state);
+
+/// Parses the Mm and Ml tags out of a bam record.
+/**
+ * @param b        BAM alignment record
+ * @param state    The base modification state pointer.
+ * @return 0 on success,
+ *         -1 on failure.
+ *
+ * This fills out the contents of the modification state, resetting the
+ * iterator location to the first sequence base.
+ */
+HTSLIB_EXPORT
+int bam_parse_basemod(const bam1_t *b, hts_base_mod_state *state);
+
+/// Returns modification status for the next base position in the query seq.
+/**
+ * @param b        BAM alignment record
+ * @param state    The base modification state pointer.
+ * @param mods     A supplied array for returning base modifications
+ * @param n_mods   The size of the mods array
+ * @return The number of modifications found on success,
+ *         -1 on failure.
+ *
+ * This is intended to be used as an iterator, with one call per location
+ * along the query sequence.
+ *
+ * If no modifications are found, the returned value is zero.
+ * If more than n_mods modifications are found, the total found is returned.
+ * Note this means the caller needs to check whether this is higher than
+ * n_mods.
+ */
+HTSLIB_EXPORT
+int bam_mods_at_next_pos(const bam1_t *b, hts_base_mod_state *state,
+                         hts_base_mod *mods, int n_mods);
+
+/// Finds the next location containing base modifications and returns them
+/**
+ * @param b        BAM alignment record
+ * @param state    The base modification state pointer.
+ * @param mods     A supplied array for returning base modifications
+ * @param n_mods   The size of the mods array
+ * @return The number of modifications found on success,
+ *         0 if no more modifications are present,
+ *         -1 on failure.
```

Comment:
bam_next_basemod does not return a -1 for a failure.

Reply:
I'm struggling to remember quite what it does do as it's been a while.

`bam_next_basemod` returns 0 if we're beyond the sequence length, which means "no more modifications are present".  Otherwise the return value is from `bam_mods_at_next_pos`.

That can return -1, but only appears to do so when pos is beyond the extent of the sequence (which I assume doesn't happen due to `bam_next_basemod` doing its own checks.

I can't really see any other way it has of generating a failure.  What sort of failure is it not spotting and reporting?

Reply:
It is more the fact that it documents that it returns a -1 and that it will never do so.

Reply:
Hmm that's an interesting philosophical question.

Checking it again I see I was wrong also.  `bam_mods_at_next_pos` is documented as returning >=0 on success and -1 on error, which it does.  `bam_next_basemod` calls this, but it has a `r > 0 ? r : 0` check so it's impossible for the error to be propagated.  This is deliberate as, right now, the only error that `bam_mods_at_next_pos` is able to return is iterating off the end of the read, which is the documented 0 return value for `bam_next_basemod`.

So you're correct that it can never return -1.

I see two options:

1.  We remove the documentation of error codes as you suggest.
2. We ignore it.

1 is correct currently, but I could see an argument for 2 as a marker for any future changes that may genuinely return an error.

Right now it's coded to deliberate avoid dynamic memory allocations, but that also gives it a maximum number of simultaneous modificationsit can report at a site (admittedly a high 256).  If we ever changed that then we'd be wanting to return an error code, which then means changing the meanings of the return values if we essentially previously documented it as being unable to return errors.

Reply:
If we treat this as the draft it is, then we can let things like this through pending future developments.

Reply:
Well, it's half returning -1 now.  Half being errors of +ve strand, but not errors of -ve strand!  I'm sure there's a better way to spot this, but I'm struggling to remember the internals except to recall it really wasn't trivial to track all the locations for both strands without computing a lot of things on-the-fly.

I tested it works on some minimal data, and I gave it a larger dataset and tried `samtools mpileup -M` which didn't report any errors, so I don't think it triggers falsely.